Empty html audit job for exploration.

DubeySandeep opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
This PR adds a one-off job for auditing empty HTML in the explorations.
Output of this one-off is in the format described below:

```
"<exp_id>:<exp_status>", { "<state_name>:interaction_id": [list(content_id of subtitledHtml having empty html.)]
```
exp_status = status of the exploration either private or public.
- We need exp_status as we can raise validation issues/warning in the private exploration and we don't have to do anything for those explorations. (still we are taking this data if in case we found that we need to do something with private explorations too while working on #6178) 

- We need interaction id because currently, the new state dict is with empty html and None interaction id.

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

DubeySandeep committed:
Added one off job to check empty html available in exploration.

DubeySandeep committed:
Updated with upstream/develop and resolved merge conflicts.

DubeySandeep committed:
Added proper docstring.

bansalnitish commented on core/domain/exp_jobs_one_off_test.py:
```diff
@@ -2695,3 +2695,126 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+        self.process_and_flush_pending_tasks()
+
+    def test_audit_job_with_empty_html_exploration_returns_content_id(self):
+
```
Add a docstring.

DubeySandeep replied:
I think test functions don't need docstring.

bansalnitish replied:
All other test functions in the file have a docstring, so just for consistency, it would be better to have one here too!

DubeySandeep replied:
Sure, I'll add one! :)

DubeySandeep replied:
Done!

bansalnitish commented on core/domain/exp_jobs_one_off_test.py:
```diff
@@ -2695,3 +2695,126 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+            'u\'Introduction:TextInput\': ['
+            'u\'content\', u\'feedback_1\', u\'hint_1\', u\'solution\']'
+            '}]'
+            ']'
```
You can move this to the previous line.

DubeySandeep replied:
Done!

bansalnitish commented on core/domain/exp_jobs_one_off_test.py:
```diff
@@ -2695,3 +2695,126 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+        self.process_and_flush_pending_tasks()
+
+    def test_audit_job_with_empty_html_exploration_returns_content_id(self):
+
```
Add a docstring.

bansalnitish commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -831,3 +831,63 @@ def reduce(key, values):
+        exploration = exp_services.get_exploration_from_model(item)
+        exp_rights = rights_manager.get_exploration_rights(item.id)
+
+        state_name_to_content_id_list = {}
```
I am not sure but would it be better to add this as a method for Exploration class in `exp_domain.py` which can be used to obtain content ids of all fields with empty html. (just as we have a [method](https://github.com/oppia/oppia/blob/develop/core/domain/exp_domain.py#L3039) defined to obtain all html strings).

It will be useful only if we will need this method for purposes other than auditing the empty html.

DubeySandeep replied:
I need content ids also the all_html_method returns html from customization agrs of few interactions. (Like Multiple choices etc.)

bansalnitish replied:
Ya, I was suggesting to have a method that returns content ids but I am not sure that it will be useful further or not.

DubeySandeep replied:
Oh, sorry, I misunderstood your last comment. Thanks for the explanation! :)

Nope, this won't be of any use as we need those content ids which have empty html, and we are planning to add validation for exploration to have "non-empty html".

bansalnitish replied:
Ah, alright then!

bansalnitish reviewed (changes requested):
Hi @DubeySandeep, Thanks for the PR. I have left a review, PTAL.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -831,3 +831,63 @@ def reduce(key, values):
+        for state_name, state in exploration.states.iteritems():
+            empty_html_content_id_list = []
+            content = state.content
+            if not (content.html).strip():
```
Don't need the parens around `content.html` here.

DubeySandeep replied:
Done!

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -831,3 +831,63 @@ def reduce(key, values):
+                    empty_html_content_id_list.append(feedback.content_id)
+
+            default_outcome = state.interaction.default_outcome
+            if default_outcome and not default_outcome.feedback.html:
```
if default_outcome is not None

DubeySandeep replied:
Done!

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -831,3 +831,63 @@ def reduce(key, values):
+                    empty_html_content_id_list.append(hint_content.content_id)
+
+            solution = state.interaction.solution
+            if solution and not solution.explanation.html:
```
Should this be `if solution is not None`?

DubeySandeep replied:
Done!

seanlip reviewed (commented):
Hi @DubeySandeep, this generally looks reasonable to me. Just to check, did you want to add a check for "RTE contains only an image tag" or did you plan to skip that?

DubeySandeep committed:
Added review changes.

DubeySandeep committed:
Resolved merge conflicts.

DubeySandeep commented:
@seanlip, I have addressed the review changes PTAL! 

> did you want to add a check for "RTE contains only an image tag" or did you plan to skip that?

We are skipping these checks (we have discussed this in detail earlier.) as there are lots of other cases for this i.e, only video, link, collapsible block present in a card.
- I think we should only care about empty HTML as users can create any meaningless content which one cannot translate.
- Maybe in the future, we can have a switch to allow users to decide whether they want their exploration to be translated? (Like parameters.)

DubeySandeep committed:
Added job to registry.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -947,3 +947,64 @@ def reduce(key, value):
+                    solution.explanation.content_id)
+            if empty_html_content_id_list:
+                if state.interaction.id is not None:
+                    key = ('%s:%s' % (state_name, state.interaction.id))
```
Use state_name.encode('utf-8') here and elsewhere

seanlip reviewed (approved):
(Feel free to merge once you've resolved the last comment!)

DubeySandeep committed:
Resolved merge conflicts and added review changes.

DubeySandeep commented:
@seanlip , I have added this job into the registry (added few more changes PTAL!). Also, I tried running this from admin page for demo exploration and I realized this is going to be a very big list:
![image](https://user-images.githubusercontent.com/16653571/52918864-36096200-3322-11e9-82bc-1229d2dd9b9d.png)



I realized we use to have empty html for default_outcome of Continue interaction.

seanlip commented:
Should we account for that case then? (Both here and in the translation
tab.).

DubeySandeep commented:
Yeah, we do care for these scenarios in the translation tab. https://github.com/oppia/oppia/blob/3c50d22f5f8eb0df5cec0b577f6aac69f33be861/core/templates/dev/head/pages/exploration_editor/translation_tab/TranslationStatusService.js#L55-L60
I think we should exclude this kinda interaction in the job result (I'll add a commit for the same) do let me know if you have any other suggestion. Thanks!

seanlip commented:
Yes, I agree that we should exclude it. Only detect the ones that we haven't accounted for yet, otherwise it'll be hard to sift through all the information.
